libvast/src/system/index.cpp

```diff
@@ -1056,144 +1211,63 @@ index(index_actor::stateful_pointer<index_state> self,
-              return;
-            }
-            auto client = caf::actor_cast<receiver_actor<atom::done>>(sender);
-            auto& midx_candidates = midx_result.partitions;
-            VAST_DEBUG("{} got initial candidates {} and from catalog {}",
-                       *self, candidates, midx_candidates);
-            candidates.insert(candidates.end(), midx_candidates.begin(),
-                              midx_candidates.end());
-            std::sort(candidates.begin(), candidates.end());
-            candidates.erase(std::unique(candidates.begin(), candidates.end()),
-                             candidates.end());
-            if (candidates.empty()) {
-              VAST_DEBUG("{} returns without result: no partitions qualify",
-                         *self);
-              self->send(self, atom::worker_v, worker);
-              rp.deliver(query_cursor{query.id, 0u, 0u});
-              self->send(client, atom::done_v);
-              return;
-            }
-            auto total = candidates.size();
-            auto scheduled = detail::narrow<uint32_t>(
-              std::min(candidates.size(), self->state.taste_partitions));
-            auto lookup = query_state{query.id, query, std::move(candidates),
-                                      std::move(worker)};
-            auto result
-              = self->state.pending.emplace(query.id, std::move(lookup));
-            VAST_ASSERT(result.second);
-            auto delta = std::chrono::steady_clock::now() - start;
-            VAST_TRACEPOINT(query_catalog, query.id.as_u64().first, total,
-                            delta.count());
-            rp.deliver(query_cursor{query.id, detail::narrow<uint32_t>(total),
-                                    scheduled});
-            // We "delegate" the first continuation back to self by spoofing
-            // the client as source. This is done so the response gets delivered
-            // to the correct recipient: the client.
-            caf::send_as(client, static_cast<index_actor>(self), query.id,
-                         scheduled);
-          },
-          [=](caf::error err) mutable {
-            VAST_ERROR("{} failed to receive candidates from catalog: {}",
-                       *self, render(err));
-            self->send(self, atom::worker_v, worker);
-            rp.deliver(std::move(err));
-          });
+        .then([=, candidates = std::move(candidates)](
+                catalog_result& midx_result) mutable {
+          auto& midx_candidates = midx_result.partitions;
+          VAST_DEBUG("{} got initial candidates {} and from meta-index {}",
+                     *self, candidates, midx_candidates);
+          candidates.insert(candidates.end(), midx_candidates.begin(),
+                            midx_candidates.end());
+          std::sort(candidates.begin(), candidates.end());
+          candidates.erase(std::unique(candidates.begin(), candidates.end()),
+                           candidates.end());
```

Comment:
There seems to be a lot of duplication with the handler ~60 lines above. Can we remove that somehow?

Reply:
the other one is now gone.